File: scripts/exporter.py
"""Export events to .ics, .csv, .json formats."""

import csv
import json
import io
from datetime import datetime
from scripts.utils import load_events, logger


def export_ics(events: list[dict] = None, month: str = None, category: str = None) -> str:
    db = load_events()
    all_events = events or db.get("events", [])

    if month:
        all_events = [e for e in all_events if e.get("start_date", "").startswith(month)]
    if category:
        all_events = [e for e in all_events if e.get("category") == category]

    lines = [
        "BEGIN:VCALENDAR",
        "VERSION:2.0",
        "PRODID:-//Big Event Calendar//EN",
        "CALSCALE:GREGORIAN",
        "METHOD:PUBLISH",
        "X-WR-CALNAME:Big Events",
        "X-WR-TIMEZONE:UTC",
    ]

    for event in all_events:
        start = _date_to_ics(event.get("start_date", ""))
        end = _date_to_ics(event.get("end_date", event.get("start_date", "")))

        lines.extend([
            "BEGIN:VEVENT",
            f"DTSTART;VALUE=DATE:{start}",
            f"DTEND;VALUE=DATE:{end}",
            f"SUMMARY:{_escape_ics(event.get('title', ''))}",
            f"DESCRIPTION:{_escape_ics(event.get('description', ''))}",
            f"CATEGORIES:{event.get('category', '').upper()}",
            f"STATUS:{_ics_status(event.get('status', ''))}",
            f"UID:{event.get('id', '')}@bigeventcalendar",
            f"LAST-MODIFIED:{_datetime_to_ics(event.get('updated_at', ''))}",
            "END:VEVENT",
        ])

    lines.append("END:VCALENDAR")
    return "\r\n".join(lines)


def export_csv(events: list[dict] = None, month: str = None, category: str = None) -> str:
    db = load_events()
    all_events = events or db.get("events", [])

    if month:
        all_events = [e for e in all_events if e.get("start_date", "").startswith(month)]
    if category:
        all_events = [e for e in all_events if e.get("category") == category]

    output = io.StringIO()
    writer = csv.writer(output)

    writer.writerow([
        "ID", "Title", "Category", "Country", "Region", "Start Date",
        "End Date", "Status", "Description", "Official Source", "Tags"
    ])

    for event in all_events:
        official = event.get("official_source", {})
        writer.writerow([
            event.get("id", ""),
            event.get("title", ""),
            event.get("category", ""),
            event.get("country", ""),
            event.get("region", ""),
            event.get("start_date", ""),
            event.get("end_date", ""),
            event.get("status", ""),
            event.get("description", "")[:200],
            official.get("url", "") if official else "",
            ", ".join(event.get("tags", [])),
        ])

    return output.getvalue()


def export_json(events: list[dict] = None, month: str = None, category: str = None, pretty: bool = True) -> str:
    db = load_events()
    all_events = events or db.get("events", [])

    if month:
        all_events = [e for e in all_events if e.get("start_date", "").startswith(month)]
    if category:
        all_events = [e for e in all_events if e.get("category") == category]

    export_data = {
        "exported_at": datetime.utcnow().isoformat(),
        "total_events": len(all_events),
        "filters": {
            "month": month,
            "category": category
        },
        "events": all_events
    }

    return json.dumps(export_data, indent=2 if pretty else None, ensure_ascii=False)


def _date_to_ics(date_str: str) -> str:
    if not date_str:
        return datetime.utcnow().strftime("%Y%m%d")
    try:
        dt = datetime.fromisoformat(date_str.replace("Z", "+00:00"))
        return dt.strftime("%Y%m%d")
    except ValueError:
        try:
            dt = datetime.strptime(date_str[:10], "%Y-%m-%d")
            return dt.strftime("%Y%m%d")
        except ValueError:
            return datetime.utcnow().strftime("%Y%m%d")


def _datetime_to_ics(dt_str: str) -> str:
    if not dt_str:
        return datetime.utcnow().strftime("%Y%m%dT%H%M%SZ")
    try:
        dt = datetime.fromisoformat(dt_str.replace("Z", "+00:00"))
        return dt.strftime("%Y%m%dT%H%M%SZ")
    except ValueError:
        return datetime.utcnow().strftime("%Y%m%dT%H%M%SZ")


def _escape_ics(text: str) -> str:
    return text.replace("\\", "\\\\").replace(";", "\\;").replace(",", "\\,").replace("\n", "\\n")


def _ics_status(status: str) -> str:
    mapping = {
        "confirmed": "CONFIRMED",
        "tentative": "TENTATIVE",
        "cancelled": "CANCELLED",
    }
    return mapping.get(status, "TENTATIVE")

File: frontend/js/exporter.js
const EventExporter = {
  exportAll(format = "ics") {
    const events = State.getFilteredEvents();
    let content, filename, mimeType;

    if (format === "ics") {
      content = this._toICS(events);
      filename = "big-events.ics";
      mimeType = "text/calendar";
    } else if (format === "csv") {
      content = this._toCSV(events);
      filename = "big-events.csv";
      mimeType = "text/csv";
    } else {
      content = JSON.stringify({ events, exported_at: new Date().toISOString() }, null, 2);
      filename = "big-events.json";
      mimeType = "application/json";
    }

    this._download(content, filename, mimeType);
  },

  exportSingle(eventId) {
    const event = State.events.find(e => e.id === eventId);
    if (!event) return;

    const content = this._toICS([event]);
    this._download(content, `${event.title || "event"}.ics`, "text/calendar");
  },

  _toICS(events) {
    const lines = [
      "BEGIN:VCALENDAR",
      "VERSION:2.0",
      "PRODID:-//Big Event Calendar//EN",
      "CALSCALE:GREGORIAN",
      "METHOD:PUBLISH",
      "X-WR-CALNAME:Big Events",
      "X-WR-TIMEZONE:UTC"
    ];

    events.forEach(event => {
      const start = this._dateToICS(event.start_date);
      const end = this._dateToICS(event.end_date || event.start_date);

      lines.push(
        "BEGIN:VEVENT",
        `DTSTART;VALUE=DATE:${start}`,
        `DTEND;VALUE=DATE:${end}`,
        `SUMMARY:${this._escapeICS(event.title || "")}`,
        `DESCRIPTION:${this._escapeICS(event.description || "")}`,
        `CATEGORIES:${(event.category || "").toUpperCase()}`,
        `STATUS:${event.status === "confirmed" ? "CONFIRMED" : "TENTATIVE"}`,
        `UID:${event.id}@bigeventcalendar`,
        "END:VEVENT"
      );
    });

    lines.push("END:VCALENDAR");
    return lines.join("\r\n");
  },

  _toCSV(events) {
    const headers = ["Date", "Title", "Category", "Country", "Status", "Description"];
    const rows = events.map(e => [
      e.start_date || "",
      `"${(e.title || "").replace(/"/g, '""')}"`,
      e.category || "",
      e.country || "global",
      e.status || "",
      `"${(e.description || "").replace(/"/g, '""').slice(0, 200)}"`
    ]);

    return [headers.join(","), ...rows.map(r => r.join(","))].join("\n");
  },

  _dateToICS(dateStr) {
    if (!dateStr) return new Date().toISOString().slice(0, 10).replace(/-/g, "");
    return dateStr.replace(/-/g, "").slice(0, 8);
  },

  _escapeICS(text) {
    return (text || "").replace(/\\/g, "\\\\").replace(/;/g, "\\;").replace(/,/g, "\\,").replace(/\n/g, "\\n");
  },

  _download(content, filename, mimeType) {
    const blob = new Blob([content], { type: mimeType });
    const url = URL.createObjectURL(blob);
    const a = document.createElement("a");
    a.href = url;
    a.download = filename;
    document.body.appendChild(a);
    a.click();
    document.body.removeChild(a);
    URL.revokeObjectURL(url);
  }
};
